File: implementations/typescript/src/ToolRegistry.ts
import { z } from 'zod';

export type ToolFunction = (args: any) => Promise<string> | string;

export interface ToolDefinition {
  type: "function";
  function: {
    name: string;
    description: string;
    parameters: any; // JSON schema
  };
}

export class ToolRegistry {
  private tools: Map<string, ToolFunction> = new Map();
  private schemas: Map<string, ToolDefinition> = new Map();

  register(name: string, description: string, func: ToolFunction, schema: any): void {
    this.tools.set(name, func);
    this.schemas.set(name, {
      type: "function",
      function: {
        name,
        description,
        parameters: schema,
      },
    });
  }

  getToolSchemas(): any[] {
    return Array.from(this.schemas.values());
  }

  async executeTool(name: string, args: any): Promise<string> {
    const func = this.tools.get(name);
    if (!func) {
      return `Error: Tool '${name}' not found.`;
    }

    try {
      const result = await func(args);
      if (typeof result !== 'string') {
        return JSON.stringify(result);
      }
      return result;
    } catch (e: any) {
      return `Error executing tool '${name}': ${e.message}`;
    }
  }
}

export const registry = new ToolRegistry();

File: implementations/fsharp/ToolRegistry.fs
namespace AgentPlatform.FSharp

open System
open System.Collections.Concurrent
open Azure.AI.OpenAI

type ToolRegistry() =
    let tools = ConcurrentDictionary<string, ToolDefinition>()

    member _.Register(name: string, description: string, handler: string -> Async<string>, parametersJson: string) =
        let def = { Name = name; Description = description; ParametersJson = parametersJson; Handler = handler }
        tools.[name] <- def

    member _.GetToolSchemas() : FunctionDefinition list =
        tools.Values
        |> Seq.map (fun t ->
            let fn = FunctionDefinition(Name = t.Name, Description = t.Description)
            if not (String.IsNullOrEmpty(t.ParametersJson)) then
                fn.Parameters <- BinaryData.FromString(t.ParametersJson)
            fn)
        |> Seq.toList

    member _.GetRegisteredNames() : string list =
        tools.Keys |> Seq.toList

    member _.ExecuteToolAsync(name: string, argsJson: string) : Async<string> =
        async {
            match tools.TryGetValue(name) with
            | true, toolDef ->
                try
                    return! toolDef.Handler argsJson
                with ex ->
                    return sprintf "Error executing tool '%s': %s" name ex.Message
            | false, _ ->
                let avail = tools.Keys |> String.concat ", "
                return sprintf "Error: Tool '%s' is not registered. Available tools: %s" name avail
        }

    /// Exposes tool execution as a composable, partially applicable function (ToolExecutor)
    member self.AsExecutor : ToolExecutor =
        fun name argsJson -> self.ExecuteToolAsync(name, argsJson)
